File: 0x10-python-network_1/103-search_twitter.py
#!/usr/bin/python3
"""
This is module 103-search_twitter.
This module tries to search on twitter
"""
import requests
import base64
import sys


def get_token(ck, cs):
    """
    retrieve token for a twitter query

    Arguments:
        ck: consumer key
        cs: consumer secret

    Returns:
       token
    """
    encoded = base64.b64encode(bytes(
        "{}:{}".format(ck, cs).encode('utf-8')))
    header = {"Authorization": "Basic " + str(encoded, encoding='utf-8'),
              "Content-Type":
              "application/x-www-form-urlencoded;charset=UTF-8"}

    r = requests.post("https://api.twitter.com/oauth2/token",
                      headers=header,
                      data={"grant_type": "client_credentials"})
    token = r.json().get("access_token")
    return (token)


if __name__ == "__main__":
    Consumer_Key = sys.argv[1]
    Consumer_Secret = sys.argv[2]

    token = get_token(Consumer_Key, Consumer_Secret)
    headers_search = {"Authorization": "Bearer " + token}

    url = "https://api.twitter.com/1.1/search/tweets.json"
    params = {'q': sys.argv[3], 'count': 5}
    r2 = requests.get(url, params=params, headers=headers_search)
    for e in r2.json().get("statuses"):
        print("[{}] {} by {}".format(
            e.get('id'), e.get('text'), e.get('user').get('name')))

File: 0x12-javascript-web_scraping/103-search_twitter.js
#!/usr/bin/node
// Use twitter search api
// allowed modules: requests, utf8 and base64
const request = require('request');
const base64 = require('base-64');
const utf8 = require('utf8');

const auth1 = process.argv[2] + ':' + process.argv[3];
const bytes = utf8.encode(auth1);
const encoded = base64.encode(bytes);
const auth = 'Basic ' + encoded;
const header = {'Authorization': auth,
  'Content-Type':
              'application/x-www-form-urlencoded;charset=UTF-8'};
const options = {
  url: 'https://api.twitter.com/oauth2/token',
  method: 'POST',
  headers: header,
  body: 'grant_type=client_credentials'
};

request(options, function (error, response, body) {
  if (!error) {
    const rJ = JSON.parse(body);
    if (rJ.token_type === 'bearer') {
      const aT = rJ.access_token;
      const headerSearch = {'Authorization': 'Bearer ' + aT};
      makeSearch(headerSearch, process.argv[4]);
    }
  }
}
);

function makeSearch (headerSearch, search) {
  const optionsSearch = {
    url: 'https://api.twitter.com/1.1/search/tweets.json',
    method: 'GET',
    headers: headerSearch,
    encoding: 'utf-8',
    qs: {'q': search, 'count': 5}
  };
  request(optionsSearch, function (error, response, body) {
    if (!error) {
      const r = JSON.parse(body);
      for (let e of r.statuses) {
        console.log('[' + e.id + '] ' + e.text + ' by ' + e.user.name);
      }
    }
  }
  );
}
